sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -954,7 +954,10 @@ object SparkSession extends Logging {
         session = new SparkSession(sparkContext, None, None, extensions, options.toMap)
         setDefaultSession(session)
         setActiveSession(session)
-        registerContextListener(sparkContext)
+        sparkContext.listenerBus.addToQueueIfNotExists(
+          new ApplicationEndListener(),
+          LiveListenerBus.SHARED_QUEUE
```

Comment:
similar here, why are we exposing this vs having a function addToSharedQueue have a parameter?

Reply:
ah, yeah. this sounds better.